galley/pkg/config/processor/transforms/serviceentry/transformer.go

```diff
@@ -0,0 +1,345 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package serviceentry
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"go.opencensus.io/tag"
+	coreV1 "k8s.io/api/core/v1"
+
+	networking "istio.io/api/networking/v1alpha3"
+	"istio.io/istio/galley/pkg/config/collection"
+	"istio.io/istio/galley/pkg/config/event"
+	"istio.io/istio/galley/pkg/config/processing"
+	"istio.io/istio/galley/pkg/config/processor/metadata"
+	"istio.io/istio/galley/pkg/config/processor/transforms/serviceentry/converter"
+	"istio.io/istio/galley/pkg/config/processor/transforms/serviceentry/pod"
+	"istio.io/istio/galley/pkg/config/resource"
+	"istio.io/istio/galley/pkg/runtime/monitoring"
+)
+
+type transformer struct {
+	options processing.ProcessorOptions
+
+	converter *converter.Instance
+
+	services  map[resource.Name]*resource.Entry
+	endpoints map[resource.Name]*resource.Entry
+	ipToName  map[string]map[resource.Name]struct{}
+
+	podHandler  event.Handler
+	nodeHandler event.Handler
+
+	handler  event.Handler
+	statsCtx context.Context
+
+	fullSyncCtr int
+
+	// The version number for the current State of the object. Every time mcpResources or versions change,
+	// the version number also change
+	version int64
+
+	// 	mcpResources map[resource.FullName]*mcp.Resource
+}
+
+var _ event.Transformer = &transformer{}
+
+// Start implements event.Transformer
+func (t *transformer) Start() {
+	t.ipToName = make(map[string]map[resource.Name]struct{})
+	t.services = make(map[resource.Name]*resource.Entry)
+	t.endpoints = make(map[resource.Name]*resource.Entry)
+
+	podCache, cacheHandler := pod.NewCache(pod.Listener{
+		PodAdded:   t.podUpdated,
+		PodUpdated: t.podUpdated,
+		PodDeleted: t.podUpdated,
+	})
+
+	t.podHandler = cacheHandler
+	t.nodeHandler = cacheHandler
+
+	t.converter = converter.New(t.options.DomainSuffix, podCache)
+
+	statsCtx, err := tag.New(context.Background(), tag.Insert(monitoring.CollectionTag,
+		metadata.IstioNetworkingV1Alpha3SyntheticServiceentries.String()))
+	if err != nil {
+		scope.Errorf("Error creating monitoring context for counting state: %v", err)
```

Comment:
Would it be reasonable to define the event.Transformer.Start method as returning an error? Otherwise, code is forced to log and can't merely propagate errors upstream.

Reply:
I had it that way originally, but I don't think it adds much value and complicates program flow quite a bit.

 Having the method not return an error forces the error case to be handled inline, as part of the runtime, which maybe solved locally via retries or optional operations.